### ISSUE TITLE
Dry-run multi-block migrations

### ISSUE THREAD

**Issue by liamaharon** [paritytech/try-runtime-cli#17]
Original issue https://github.com/paritytech/substrate/issues/14291

--- 

              @kianenigma am I right to assume we'll want to make it easy to test these with try-runtime-cli? Ideally should be as easy as `on-runtime-upgrade`, maybe make it auto-mine blocks until all multi-block migrations are complete & provide a hook that runs afterwards?

_Originally posted by @liamaharon in https://github.com/paritytech/substrate/issues/14275#issuecomment-1570824518_
            

Kian response:
> Good to think about it early, and yes I think it is possible with little effort. I believe follow-chain will already do everything needed to test an MBM?

---

I don't think `follow-chain` is necessarily a good solution. 
1. We need to call each upgrade's `post_upgrade` hook once its migration has finished executing.
2. We can't use snapshots with `follow-chain`, making development iterations much slower and requiring a live node running. Some of the plumbing in `fast-forward` is might be a better solution here?

Ideally, I think we'd be able to run `on-runtime-upgrade` like usual and it'll fast-forward until all migrations have completed, and then run a post_migration hook. Unfortunately fast-forward is not production ready due to the [inherents issue](https://github.com/paritytech/substrate/issues/13887), it seems this feature makes fixing the inherents issue much higher priority now?

**Comment by liamaharon** [comment]
The inherents issue is fixed, so this is unblocked :)

**Comment by liamaharon** [comment]
There're two options to consider to add MBM testing to `try-runtime`, I'd like to come to some kind of consensus before starting work on one.

## 1. Run MBMs in existing `try_on_runtime_upgrade` logic until complete

This approach would integrate MBMs into the existing `try_on_runtime_upgrade` logic, so they're tested as part of the existing flow.

Simply, after dry-running the migrations, `step` would be called on all MBMs until completion.

Pros:
- Very simple to implement and understand logic
- Non-breaking change to `try-runtime-cli`

Cons: 
- Less "realistic" given the MBMs are progressing without actual blocks being mined.

## 2. Continuously mine new blocks after `try_on_runtime_upgrade` has executed

In this approach, we'd continue to mine empty blocks after the runtime upgrade, checking the MBM status (maybe needs a new runtime api?) each block to check the state of MBMs.

Pros:
- More realistic, since inherents will be applied and the block number will progress between each MBM step as it would when run for real

Cons:
- More complex implementation
- Requires new runtime api (I think)
- Breaking change to `try-runtime-cli`, and `try-runtime-cli` will become slightly more difficult to use because the user will need to specify their chain inherents when dry-running their migrations `on-runtime-upgrade`.

## My thoughts

I'm on the fence with this one, but slightly leaning towards option 1.

Although option 2 is slightly more realistic, I can't foresee why MBMs would be built in such a way that inherents should impact their progress --- that sounds like it would be an anti-pattern anyway?

If option 2 did not come with the drawbacks of being significantly more complex to implement, or making `try-runtime-cli` more complex to use for the end-user, I would suggest it. But given the drawbacks, my gut feeling is it's not worth the additional time or complexity to implement and we should proceed with option 1.

**Comment by ggwpez** [comment]
Yea i think initially we can do option 1, to get something going quick.

> Although option 2 is more realistic, I can't foresee why MBMs would be built in such a way that inherents should impact their progress. That sounds like it would be an anti-pattern anyway, that would be fine to not allow?

"should not" and "do not" is exactly why we need to test it. Steaming some data in from an inherent to a MBM would work i guess - but very non-standard to do.  
Would it work to add a `--mbm-test-strategy` flag or something that can toggle between option 1 and (possibly future) option 2? Not sure how difficult it would be to write it in a general way to have this possibility.

**Comment by kianenigma** [comment]
My main reservation towards option 1 is that it might miss something like checking the weight of each step. Can we strictly ensure that a dry-run-ed MBM with approach 1 will detect if it is overweight? 

If these two are the only options, I would also lean towards 1. I agree that getting something out the door to test MBMs are more important. But I don't like option 2 per-se as I think it is not the right abstraction. 

If try-runtime provides a good system to expresses "please dry-run 5 empty blocks for me", which was always part of the plan, testing MBMs is more or less trivial. I think this is what we should aim for, and then testing all types of migrations will be a specialized version of this. 

In other words, if try-runtime can create empty blocks for a given, which I believe with the correct inherent extension it can, even the existing `try-runtime-upgrade` would be rendered deprecated, as it can be tested more realistically by just creating the next empty block.  

Even now, I believe the main advantage of `try_runtime_upgrade` being a custom runtime-api is such that it can bypass block level details such as inherents and just execute the migration code path.

**Comment by liamaharon** [comment]
It seems there is consensus that we should pursue option 2, even if it will take longer.

My first question about implementing it is what is the best way to get the MBM migration status from try-runtime-cli while mining empty blocks?

Initially I was thinking to directly query the event storage, but I don't think I'd be able to decode it because the `RuntimeEvent` type varies between every runtime, and we don't know what it is.

We also I don't think can query the pallet storage directly, because the key it's under varies across runtimes.

We could introduce a new runtime api for querying MBM status, but that's far less than ideal because it creates non-trivial work for runtime devs to implement.

The approach that seems best would be to modify `Executive`'s `try_execute_block` methods return value to include information about MBM status. Then, looks like `try-runtime-cli` can get MBM information from fn it's calling already.  

@bkchr @ggwpez @kianenigma let me know if this sounds sensible.

**Comment by kianenigma** [comment]
> My first question about implementing it is what is the best way to get the MBM migration status from try-runtime-cli while mining empty blocks?

I think this new return type of `initailize_block` is exactly what is meant for this: 

https://paritytech.github.io/polkadot-sdk/master/sp_runtime/enum.ExtrinsicInclusionMode.html

I think if you update this code to respect this new type, you would already have MBM testing done: 

https://github.com/paritytech/try-runtime-cli/blob/226514ef508f1f18cc5b84a3c7a71e5e06a24725/core/src/commands/fast_forward.rs#L153-L222

**Comment by ggwpez** [comment]
I think just for PoC, we can also use chopsticks, since it exposes a block-mining RPC and the possibility to decode events.

>I think this new return type of initailize_block is exactly what is meant for this:

Kind of, yes. Although there is a hail-mary last resort option, where a failed migration can decide what to do. In that case, one option is `ForceUnstuck`, and then the chain resumes. So *just* looking at that could be misinterpreted as success.  
It is still better to either 1) parse events or 2) create a Runtime API.

**Comment by liamaharon** [comment]
> I think if you update this code to respect this new type, you would already have MBM testing done:

Unfortunately as @ggwpez said just checking the `ExtrinsicInclusionMode` won't give us any solid MBM status.

> I think just for PoC, we can also use chopsticks, since it exposes a block-mining RPC and the possibility to decode events.

I am worried devex will suck if we start requiring two completely different tools to test a runtime upgrade. Also, I'm not sure Chopsticks handles pre/post hooks.

For now proceed with modifying the `try_execute_block` ret value unless u have some other ideas.

**Comment by bkchr** [comment]
> Unfortunately as @ggwpez said just checking the `ExtrinsicInclusionMode` won't give us any solid MBM status.

What solid MBM status? What do you need to know besides if they are still running?

**Comment by liamaharon** [comment]
> > Unfortunately as @ggwpez said just checking the `ExtrinsicInclusionMode` won't give us any solid MBM status.
> 
> What solid MBM status? What do you need to know besides if they are still running?

Need to also know whether or not it succeeded.

**Comment by bkchr** [comment]
Hmm? Run `post_runtime_upgrade` after all steps are finished.

**Comment by liamaharon** [comment]
> Hmm? Run `post_runtime_upgrade` after all steps are finished.

Do you mean some try-runtime feature-gated code that will panic if the MBMs fail?

**Comment by bkchr** [comment]
No I mean as we have for `OnRuntimeUpgrade`  the pre and post runtime upgrade functions. Here we could have something similar, aka a function that we call after the entire MBM is finished.
